tests/periph_uart_mode/main.c

```diff
@@ -0,0 +1,153 @@
+ *
+ * @}
+ */
+
+
+
+/*
+ * This application is intended for testing the UART mode function
+ * on CPUs that only have 1 UART. The existing UART test does not
+ * switching UART configuration mode on the device being used for
+ * shell. This test will cycle through all permutations of UART modes
+ * and print the mode to STDOUT. The test is evaluated by hooking up
+ * a probe or logic analyzer and confirming that each string can be
+ * read correctly. The strings will be printed as such:
+ * <data-bits><parity-bits><stop-bits>
+ * Example: 6O1 would be 6 data bits, odd parity, 1 stop bit
+ * If a mode is not supported, then a message indicating that will
+ * be printed using configuration from the last mode);
+ */
+
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "periph/uart.h"
+#include "xtimer.h"
+
+/* UART_DEV is always 0 since this test is for device with 1 UART */
+#define _UART_DEV (UART_DEV(0))
+
+
+static void _print_status(int8_t status, uart_data_bits_t data_bits,
+                          uart_parity_t parity, uart_stop_bits_t stop_bits) {
+    char mode[4];
+
+    /* If status < 0, print a header indicating an error */
+    if (status < 0) {
+        printf("Unsupported Mode: ");
+    }
+
+    switch (data_bits) {
+        case UART_DATA_BITS_5:
+            mode[0] = '5';
+            break;
+        case UART_DATA_BITS_6:
+            mode[0] = '6';
+            break;
+        case UART_DATA_BITS_7:
+            mode[0] = '7';
+            break;
+        case UART_DATA_BITS_8:
+            mode[0] = '8';
+            break;
+        default:
+            break;
+    }
+
+    switch (parity) {
+        case UART_PARITY_NONE:
+            mode[1] = 'N';
+            break;
+        case UART_PARITY_EVEN:
+            mode[1] = 'E';
+            break;
+        case UART_PARITY_ODD:
+            mode[1] = 'O';
+            break;
+        case UART_PARITY_MARK:
+            mode[1] = 'M';
+            break;
+        case UART_PARITY_SPACE:
+            mode[1] = 'S';
+            break;
+        default:
+            break;
+    }
+
+    switch (stop_bits) {
+        case UART_STOP_BITS_1:
+            mode[2] = '1';
+            break;
+        case UART_STOP_BITS_2:
+            mode[2] = '2';
+            break;
+        default:
+            break;
+    }
+
+    mode[3] = '\0';
+
+    printf("%s\n", mode);
+}
+
+
+int main(void)
+{
+
+    /* Use arrays so that each option can be easily iterated over */
+    uart_data_bits_t data_bits[4] = {
+        UART_DATA_BITS_5,
+        UART_DATA_BITS_6,
+        UART_DATA_BITS_7,
+        UART_DATA_BITS_8
+    };
+
+    uart_parity_t parity[5] = {
+        UART_PARITY_NONE,
+        UART_PARITY_EVEN,
+        UART_PARITY_ODD,
+        UART_PARITY_MARK,
+        UART_PARITY_SPACE
+    };
+
+    uart_stop_bits_t stop_bits[2] = {
+        UART_STOP_BITS_1,
+        UART_STOP_BITS_2
+    };
+
+    /* Test each permutation */
```

Comment:
If you could print the supported values in the beginning that would be nice before any changes to the mode.

Reply:
So cycle through each and mode and aggregate the results, print them, then cycle through again and print each mode? I don't think there's any way to know which values are supported before calling the mode switch function.

Reply:
I think it's important to actually have the user verify that the mode switches caused the data to be written with the new configuration, and not just that uart_mode returned the correct status code. I think the only good way to do that is to print something with each mode

Reply:
Then save it and put it at the end or reiterate at the end.  We want an easy trigger setup on the scope.

Reply:
OK, let me know what you think about that. Not sure what you mean by "easy trigger setup", I still don't have a good way to mark each mode (except for the sleep), but this prints out all results with default mode configuration at the end.

Reply:
By easy trigger I mean if I have output the scope data at the start and the console data after, this way I just can trigger on the first edge instead of searching through the console output.

But very nice!

Reply:
ah, I see, that makes sense.